### DISCUSSION TITLE
Conditional toolbar buttons

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#5476] [Q&A]
I have a custom toolbar that includes of CoreTools toolbar buttons (pls see code below)

```
    const myToolbar: CommonToolbarItem[] = [
      ToolbarHelper.createToolbarItemFromItemDef(10, CoreTools.measureToolGroup ),
      ToolbarHelper.createToolbarItemFromItemDef(20, CoreTools.sectionToolGroupWithPanel)
     ]

```
I would like to add two conditional button (using ```isHidden``` parameter), one to clear measurement and second to clear section after user selects one of these tools. 

I want to know if there is SyncUiEvent that I can listen to find out when user selects one of the menu options in these two groups.  It seems a ```toolactivated``` event is generated but it doesn't seem to indicate which of these two were chosen.

Thanks,

Ovais

**Comment by NancyMcCallB** [comment]
Sorry, @ovaisa ! I forgot to come back here to answer. Take a look at https://github.com/iTwin/appui/blob/master/ui/appui-react/src/appui-react/selection/HideIsolateEmphasizeManager.ts to see how to dispatch a custom SyncUi event. If it's not what you need, let me know.

Nancy McCall

**Comment by ovaisa** [comment]
Hi Nancy,

I am already using custom SyncUi events extensively to manage toolbars buttons/appearance (using ```ConditionalStringValue```).   My question is related to UIEvents generated by iTwin CoreTools, specifically the ones in the code above. 

If I have my own callback for toolbar buttons, I know when a button is selected and I can generate UIEvents.  But in case of CoreTools buttons, actions are managed by iTwin and I am unable to find out when user selects these CoreTools buttons.  In this specific instance, I want to know when user selects the two CoreTools buttons in the code above (```CoreTools.measureToolGroup``` and ```CoreTools.sectionToolGroupWithPanel```).

In an unrelated post much earlier, I was told by Bentley SyncUi events are generated during CoreTool actions.  I want to know what they are so that I can react to them to manage toolbar appearance (for these two CoreTools).

Ovais

**Reply by NancyMcCallB** [reply]
Ah! I misunderstood. 

Yes, the ToolAdmin dispatches a refresh syncUI event for tools. You can find the details of the implementation here: https://github.com/iTwin/itwinjs-core/blob/66a52391cd069ad7ef156cf8a7c112e4532dafc1/core/frontend/src/tools/ToolAdmin.ts#L1673

"tool-admin-refresh-ui"-<toolId> is probably what you're looking for.

**Comment by ovaisa** [comment]
I added the following code to examine what events are being raised

```
  public static initSyncUI = () => {
    SyncUiEventDispatcher.onSyncUiEvent.addListener(this.handleSyncUiEvent);
  }
  private static handleSyncUiEvent = async (args: UiSyncEventArgs) => {
    console.log("SyncUI: ", args.eventIds);
 }
```

When I select one of the buttons added by ```measureToolGroup```, I get the following output:

SyncUI:  Set(1) {'configurableui:set_toolprompt'}
SyncUI:  Set(1) {'configurableui:set_toolprompt'}
SyncUI:  Set(1) {'toolactivated'}

I don't see "tool-admin-refresh-ui" event.  Am I missing something.  In addition, I would like to find out exactly which tool ran, not just any tool, so that I can update toolbar accordingly.

Thanks,

Ovais

**Reply by NancyMcCallB** [reply]
You are correct -- the measure tools do not dispatch the sync events. @bbastings, do the core tools generate any tool-specific events that an app can listen for?
